pandas/tests/reshape/test_concat.py

```diff
@@ -1771,6 +1772,148 @@ def test_concat_tz_series_tzlocal(self):
+        result = pd.concat([first, second], axis=1)
+        assert_frame_equal(result, expected, check_datetimelike_compat=True)
+
+        # both sides timezone-aware
+        second = pd.DataFrame([[pd.NaT]], dtype=dtype, columns=[1])
+
+        # upcasts to tz-aware
+        expected = expected.apply(lambda x: x.astype(dtype))
+        result = pd.concat([first, second], axis=1)
+        assert_frame_equal(result, expected, check_datetimelike_compat=True)
+
+    def test_concat_NaT_dataframes_mixed_timestamps_and_NaT(self):
+        # GH 12396
+
+        # non-timezone aware
+        first = pd.DataFrame([[pd.NaT], [pd.NaT]])
+        second = pd.DataFrame([[pd.Timestamp('2015/01/01')],
+                               [pd.Timestamp('2016/01/01')]])
+
+        expected = pd.DataFrame([pd.NaT, pd.NaT, second.iloc[0, 0],
+                                second.iloc[1, 0]], index=[0, 1, 0, 1])
+
+        result = pd.concat([first, second], axis=0)
+        assert_frame_equal(result, expected, check_datetimelike_compat=True)
+
+        # one side timezone-aware
+        dtype = DatetimeTZDtype('ns', tz='UTC')
+        second = second.apply(lambda x: x.astype(dtype))
+
+        result = pd.concat([first, second], axis=0)
+        expected = expected.apply(lambda x: x.astype(dtype))
+        assert_frame_equal(result, expected, check_datetimelike_compat=True)
+
+    def test_concat_NaT_series_dataframe_all_NaT(self):
+        # GH 12396
+
+        # non-timezone aware
+        first = pd.Series([pd.NaT, pd.NaT])
+        second = pd.DataFrame([[pd.Timestamp('2015/01/01')],
+                               [pd.Timestamp('2016/01/01')]])
+
+        expected = pd.DataFrame([pd.NaT, pd.NaT, second.iloc[0, 0],
+                                 second.iloc[1, 0]], index=[0, 1, 0, 1])
+
+        result = pd.concat([first, second])
+        assert_frame_equal(result, expected, check_datetimelike_compat=True)
+
+        # one side timezone-aware
+        dtype = DatetimeTZDtype('ns', tz='UTC')
+        second = second.apply(lambda x: x.astype(dtype))
+
+        result = pd.concat([first, second])
+
+        expected = expected.apply(lambda x: x.astype(dtype))
+        assert_frame_equal(result, expected, check_datetimelike_compat=True)
+
+        # both sides timezone-aware
+        first = first.astype(dtype)
+        result = pd.concat([first, second])
+        assert_frame_equal(result, expected, check_datetimelike_compat=True)
+
+        first = pd.DataFrame([[pd.NaT], [pd.NaT]])
+        second = pd.DataFrame([[pd.Timestamp('2015/01/01', tz='UTC')],
+                               [pd.Timestamp('2016/01/01', tz='US/Eastern')]])
+
+        expect = pd.DataFrame([pd.NaT, pd.NaT, second.iloc[0, 0],
+                               second.iloc[1, 0]], index=[0, 1, 0, 1])
+
+        result = pd.concat([first, second], axis=0)
+        assert_frame_equal(result, expect, check_datetimelike_compat=True)
+
+    def test_concat_empty_datetime_series(self):
```

Comment:
see if there are other very similar tests in this file, if so move them next to these (or put these tests next to them). we have to split this file as its getting too big anyhow.